test/irmin/test_tree.ml

```diff
@@ -74,14 +82,107 @@ let test_diff _ () =
         "Changed metadata"
         [ ([ "k" ], `Updated (("v", Left), ("v", Right))) ]
 
+(* Correct stats for a completely lazy tree *)
+let lazy_stats = Tree.{ nodes = 0; leafs = 0; skips = 1; depth = 0; width = 0 }
+
 let test_clear _ () =
-  let x = List.init 830829 (fun i -> string_of_int i) in
-  Lwt_list.fold_left_s (fun acc i -> Tree.add acc [ i ] i) Tree.empty x
-  >|= fun large_tree -> Tree.clear large_tree
+  let size = 830829 in
+  let* large_tree =
+    List.init size string_of_int
+    |> Lwt_list.fold_left_s (fun acc i -> Tree.add acc [ i ] i) Tree.empty
+  in
+  let* () =
+    Tree.stats ~force:false large_tree
+    >|= Alcotest.(gcheck Tree.stats_t)
+          "Before clear, root node is eagerly evaluated"
+          { nodes = 1; leafs = size; skips = 0; depth = 1; width = size }
+  in
+  Tree.clear large_tree;
+  let* () =
+    Tree.stats ~force:false large_tree
+    >|= Alcotest.(gcheck Tree.stats_t)
+          "After clear, root node is no longer cached" lazy_stats
```

Comment:
it's unclear to me why this returns `lazy_stats`: the tree is not saved to disk yet, it should have kept everything in memory? Can you still re-read the contents of that tree after a clear (I think we should expect this work)?

Reply:
The `clear` wipes out the cache held in `Node.t.info`, but not the actual value – it's indeed still possible to read the contents of the tree afterwards, that just requires a forcing action.

```
- Node info before:

  { value = None;
    map =
    Some ([("9", Contents (Value ("9"))); ("8", Contents (Value ("8")));
            ("7", Contents (Value ("7"))); ("6", Contents (Value ("6")));
            ("5", Contents (Value ("5"))); ("4", Contents (Value ("4")));
            ("3", Contents (Value ("3"))); ("2", Contents (Value ("2")));
            ("19", Contents (Value ("19"))); ("18", Contents (Value ("18")));
            ("17", Contents (Value ("17"))); ("16", Contents (Value ("16")));
            ("15", Contents (Value ("15"))); ("14", Contents (Value ("14")));
            ("13", Contents (Value ("13"))); ("12", Contents (Value ("12")));
            ("11", Contents (Value ("11"))); ("10", Contents (Value ("10")));
            ("1", Contents (Value ("1"))); ("0", Contents (Value ("0")))
            ...]);
    hash = None;
    findv_cache = None }

- Calling Node.clear
- Node info after:

  { value = None;
    map = None;
    hash = None;
    findv_cache = None }
```

Reply:
hum that behaviour is a bit weird - I think we want fold to iterate over the value even if it has an empty cache.